packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -303,24 +335,55 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
         // note: this check throws an error if the object is found but does not exist in this namespace
-        const namespaceString = SavedObjectsUtils.namespaceIdToString(namespace);
-        const [{ error, existingDocument }] = await preflightCheckForCreate({
-          registry: this._registry,
-          client: this.client,
-          serializer: this._serializer,
-          getIndexForType: this.getIndexForType.bind(this),
-          createPointInTimeFinder: this.createPointInTimeFinder.bind(this),
-          objects: [{ type, id, overwrite, namespaces: initialNamespaces ?? [namespaceString] }],
-        });
-        if (error) {
-          throw SavedObjectsErrorHelpers.createConflictError(type, id);
-        }
-        savedObjectNamespaces =
-          initialNamespaces || getSavedObjectNamespaces(namespace, existingDocument);
-        existingOriginId = existingDocument?._source?.originId;
-      } else {
-        savedObjectNamespaces = initialNamespaces || getSavedObjectNamespaces(namespace);
+        preflightResult = (
+          await preflightCheckForCreate({
+            registry: this._registry,
+            client: this.client,
+            serializer: this._serializer,
+            getIndexForType: this.getIndexForType.bind(this),
+            createPointInTimeFinder: this.createPointInTimeFinder.bind(this),
+            objects: [{ type, id, overwrite, namespaces: initialNamespaces ?? [namespaceString] }],
+          })
+        )[0];
       }
+      savedObjectNamespaces =
+        initialNamespaces || getSavedObjectNamespaces(namespace, preflightResult?.existingDocument);
+      existingOriginId = preflightResult?.existingDocument?._source?.originId;
+    }
+
+    const spacesToEnforce = new Set(initialNamespaces).add(namespaceString); // Always check/enforce authZ for the active space
+    const existingNamespaces = preflightResult?.existingDocument?._source?.namespaces || [];
+    const spacesToAuthorize = new Set(existingNamespaces);
+    spacesToAuthorize.delete(ALL_NAMESPACES_STRING); // Don't accidentally check for global privileges when the object exists in '*'
+    const authorizationResult = await this._securityExtension?.checkAuthorization({
+      types: new Set([type]),
+      spaces: new Set([...spacesToEnforce, ...spacesToAuthorize]), // existing namespaces are included so we can later redact if necessary
+      actions: ['create'],
+      // If a user tries to create an object with `initialNamespaces: ['*']`, they need to have 'create' privileges for the Global Resource
+      // (e.g., All privileges for All Spaces).
+      // Inversely, if a user tries to overwrite an object that already exists in '*', they don't need to 'create' privileges for the Global
+      // Resource, so in that case we have to filter out that string from spacesToAuthorize (because `allowGlobalResource: true` is used
+      // below.)
```

Comment:
It feels like this is a lot of auth logic that's bleeding into the repository. Is there a way we could move this into securityExtension? I didn't have time to investigate if it's possible, but in theory we could just give all the information to the securityExtension and let it do all of this logic?